tun/tun_illumos.go

```diff
@@ -0,0 +1,355 @@
+		}
+		ifindex := int(*(*uint16)(unsafe.Pointer(&data[12 /* ifindex */])))
+		if ifindex != tunIfindex {
+			continue
+		}
+
+		iface, err := net.InterfaceByIndex(ifindex)
+		if err != nil {
+			//tun.errors <- err
+			//return
+			// XXX returning this error breaks everything??
+			continue
+		}
+
+		// Up / Down event
+		up := (iface.Flags & net.FlagUp) != 0
+		if up != statusUp && up {
+			tun.events <- EventUp
+		}
+		if up != statusUp && !up {
+			tun.events <- EventDown
+		}
+		statusUp = up
+
+		// MTU changes
+		if iface.MTU != statusMTU {
+			tun.events <- EventMTUUpdate
+		}
+		statusMTU = iface.MTU
+	}
+}
+
+func tunDestroy(name string) error {
+	var l unix.Lifreq
+	if err := l.SetName(name); err != nil {
+		return err
+	}
+	for _, proto := range []int{unix.AF_INET, unix.AF_INET6} {
+		fd, err := unix.Socket(proto, unix.SOCK_DGRAM, 0)
+		if err != nil {
+			return fmt.Errorf("could not open Socket of type %v", proto)
+		}
+
+		// illumos defines the ioctl number as a signed int, but the
+		// common x/sys/unix functions use uint. Hand-cast things to help
+		// the compiler figure it out.
+		reqnum := int(unix.SIOCGLIFMUXID)
+		if err := unix.IoctlLifreq(fd, uint(reqnum), &l); err != nil {
+			continue
+		}
+
+		id := l.GetLifruInt()
+
+		if err := unix.IoctlSetInt(fd, unix.I_PUNLINK, id); err != nil {
+			continue
+		}
+
+		unix.Close(fd)
+	}
+	return nil
+}
+
+func CreateTUN(name string, mtu int) (Device, error) {
+	if name != "tun" {
+		return nil, fmt.Errorf("Interface name must be 'tun'")
+	}
+
+	fd, err := unix.Open(tunNode, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("Could not open TUN (%s)", tunNode)
+	}
+
+	// Ask the TUN driver for a new PPA number:
+	var ppa int
+	var s unix.Strioctl
+	s.Cmd = unix.TUNNEWPPA
+	for try := 0; try < 128; try++ {
+		s.SetInt(try)
+		ppa, err = unix.IoctlSetStrioctlRetInt(fd, unix.I_STR, &s)
+		if err != nil {
+			if err == unix.EEXIST {
+				// PPA already in use, try the next one.
+				continue
+			}
+			unix.Close(fd)
+			return nil, err
+		}
+		break
+	}
+
+	assignedName := fmt.Sprintf("tun%d", ppa)
+	// Immediately do a full cleanup of anything stale left behind by a previous crash
+	tunDestroy(assignedName)
+
+	dgramFD, err := unix.Socket(unix.AF_INET, unix.SOCK_DGRAM, 0)
+	if err != nil {
+		return nil, fmt.Errorf("could not open a UDP Socket")
+	}
+	defer unix.Close(dgramFD)
+
+	// Unclear why, but we need to open a second handle on the tun device for
+	// the ioctls that follow. Reusing the existing one does not work.
+	tunFD, err := unix.Open(tunNode, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("Could not open TUN (%s)", tunNode)
+	}
+	defer unix.Close(tunFD)
+
+	// Push the IP module onto the new TUN device.
+	if err = unix.IoctlSetString(tunFD, unix.I_PUSH, "ip"); err != nil {
+		unix.Close(fd)
+		tunDestroy(assignedName)
+		return nil, fmt.Errorf("unable to push IP module: %v", err)
+	}
+
+	// illumos defines the ioctl number as a signed int, but the
```

Comment:
No need to repeat this comment each time, once on first use of the pattern is enough. Same comment throughout this file.